s3upload.js

```diff
@@ -32,16 +37,19 @@ function S3Upload(options) {
             this[option] = options[option];
         }
     }
-    var files = this.fileElement ? this.fileElement.files : this.files || [];
-    this.handleFileSelect(files);
+    this.handleFileSelect(this.files);
 }
 
 S3Upload.prototype.handleFileSelect = function(files) {
     var result = [];
+    var that = this;
+    var next = function (file) {
+        that.onProgress(0, 'Waiting', file);
+        result.push(that.uploadFile(file));
+    }
     for (var i=0; i < files.length; i++) {
         var file = files[i];
-        this.onProgress(0, 'Waiting', file);
-        result.push(this.uploadFile(file));
+        this.onBeforeUpload(file, next)
```

Comment:
The idea is that the file passed back in next(file) might not be the same file at all.  This was my application where I'm actually passing back a Blob that I added a "name" attribute to so the s3Upload code works.  xhr.send is totally happy either way and instead of uploading the original file I'm uploading an image I created, using the original file as a starting point..